### DISCUSSION TITLE
How to improve the concurrent write limit due to SQLite?

### DISCUSSION THREAD

**Discussion by kawpii** [pocketbase/pocketbase#5524] [Q&A] [answered]
As stated in https://sqlite.org/whentouse.html

> Many concurrent writers? → choose client/server
> 
> If many threads and/or processes need to write the database at the same instant (and they cannot queue up and take turns) then it is best to select a database engine that supports that capability, which always means a client/server database engine.
> 
> SQLite only supports one writer at a time per database file. But in most cases, a write transaction only takes milliseconds and so multiple writers can simply take turns. SQLite will handle more write concurrency than many people suspect. Nevertheless, client/server database systems, because they have a long-running server process at hand to coordinate access, can usually handle far more write concurrency than SQLite ever will.

Is not this important in realtime feature? Is WAL only improvement available?

**Comment by ganigeorgiev** [comment]
I don't understand what is the question, nor what is the actual problem.

If the only concern is "performance", then you could check https://github.com/pocketbase/benchmarks.

More than this I don't think I can help further and I find discussions about performance pointless if there is nothing concrete to work/discuss upon.

**Reply by kawpii** [reply]
Problem is there can be only single writer in SQLite according to the quoted text, so in a chat application where multiple users send messages (writes) it creates a bottle neck.

**Reply by ganigeorgiev** [reply, marked as answer]
Multiple writes just queue up and as mentioned in the above quoted SQLite doc, "SQLite will handle more write concurrency than many people suspect".

Please explore the benchmark results. If handling 50k create record operations over the web API in ~1min on the cheapest shared vCPU Hetzner VPS is not OK for your use case then simply use something else.
